### PR TITLE
1.10/1566 flexibility of voucher amounts in different areas eee vvv

### PR DESCRIPTION
**Pull request by charlesstrange2**
https://trello.com/c/1N2OTi1H/1566-flexibility-of-voucher-amounts-in-different-areas-eee-vvv
Sadly, does not contain 
- dynamic updates to the rules text on the edit page
- a migration to set the live value for SK
- variable values tests.

### CODE REVIEW

File: app/Services/VoucherEvaluator/EvaluatorFactory.php [hunk 1]
```diff
@@ -51,106 +38,69 @@ public static function make($type = null, Carbon $offsetDate = null)
...
                 'disqualifiers' => [
-                    new ChildIsSecondarySchoolAge($offsetDate)
+                    "ChildIsPrimarySchoolAge" => new ChildIsPrimarySchoolAge($offsetDate, 0),
                 ],
             ],
-            Family::class => [
+            "App\Family" => [
                 'credits' => [
-                    new FamilyIsPregnant(null, 3)
+                    "FamilyIsPregnant" => new FamilyIsPregnant($offsetDate, 3)
```
Comment (AlinChindea) [inline app/Services/VoucherEvaluator/EvaluatorFactory.php:72]: doesn't a pregnant family need a notice about an upcoming birth?

File: app/Services/VoucherEvaluator/EvaluatorFactory.php [hunk 2]
```diff
@@ -51,106 +38,69 @@ public static function make($type = null, Carbon $offsetDate = null)
...
                 'credits' => [],
                 'notices' => [],
                 'relations' => ['family'],
             ],
         ];
+        $namespace = "App\Services\VoucherEvaluator\Evaluations";
+        // Iterate over the modEvaluations and replace/add them
+        foreach ($modEvaluations as $mod) {
+            $className = $namespace . '\\' . $mod->name;
+            // Check we can
```
Comment (AlinChindea) [inline app/Services/VoucherEvaluator/EvaluatorFactory.php:89]: unfinished comment? 🤔

File: database/factories/ModelFactory.php [hunk 3]
```diff
@@ -479,7 +467,7 @@
 });
 
 // Child - over School Age
-$factory->defineAs(App\Child::class, 'overSchoolAge', function (Faker\Generator $faker) {
+$factory->defineAs(App\Child::class, 'isPrimarySchoolAge', function (Faker\Generator $faker) {
```
Comment (AlinChindea) [inline database/factories/ModelFactory.php:470]: I will have to change the test seeds I've created to reflect these changes or we can do so in this PR maybe...

File: app/Services/VoucherEvaluator/EvaluatorFactory.php (hunk 1 above)
Comment (charlesstrange2) [reply app/Services/VoucherEvaluator/EvaluatorFactory.php:72]: No, we found that could cause issues if the child doesn't get born for some reason...

File: app/Services/VoucherEvaluator/EvaluatorFactory.php (hunk 2 above)
Comment (charlesstrange2) [reply app/Services/VoucherEvaluator/EvaluatorFactory.php:89]: will fix

File: app/Services/VoucherEvaluator/Evaluations/ChildIsBetweenOneAndSecondarySchoolAge.php [hunk 4]
```diff
@@ -6,19 +6,30 @@
 use App\Specifications\IsUnderStartDate;
 use Carbon\Carbon;
 use Chalcedonyt\Specification\AndSpec;
+use Chalcedonyt\Specification\NotSpec;
 
-class ChildIsUnderSecondarySchoolAge extends BaseChildEvaluation
+class ChildIsBetweenOneAndSecondarySchoolAge extends BaseChildEvaluation
```
Comment (AlinChindea) [inline app/Services/VoucherEvaluator/Evaluations/ChildIsBetweenOneAndSecondarySchoolAge.php:11]: You probably don't need this as you got it covered by `ChildIsBetweenOneAndPrimarySchoolAge` and `ChildIsPrimarySchoolAge`. but like 🎅 , check it twice 👯‍♀️ 👯‍♂️

File: database/seeds/SponsorsSeeder.php [hunk 5]
```diff
@@ -11,11 +12,93 @@ class SponsorsSeeder extends Seeder
...
+        // only gets verifications
+        // used by 6MAY-20-VC1-CH2-032020
+        $sponsor4 = App\Sponsor::find(4);
+        $sponsor4->evaluations()->saveMany($this->veryfiesKids());
+
+        // 4MAY20-VC1-CH1P-HA-012020 is in sponsor 5, unmodified
+        // 1MAY20a-VC2-CH1-HI-042019 is in sponsor 5, unmodified
+    }
+
+    public function veryfiesKids()
```
Comment (katjam) [inline database/seeds/SponsorsSeeder.php:47]: Is this spelled right?

File: app/Services/VoucherEvaluator/EvaluatorFactory.php [hunk 6]
```diff
@@ -51,106 +38,69 @@ public static function make($type = null, Carbon $offsetDate = null)
...
-
-                    // ... or if almost secondary school age
-                    new ChildIsAlmostSecondarySchoolAge($offsetDate),
+                    "ChildIsAlmostOne" => new ChildIsAlmostOne($offsetDate, 0),
+                    "ChildIsAlmostPrimarySchoolAge" => new ChildIsAlmostPrimarySchoolAge($offsetDate, 0),
                 ],
                 'relations' => [],
                 'disqualifiers' => [
-                    new ChildIsSecondarySchoolAge($offsetDate)
+                    "ChildIsPrimarySchoolAge" => new ChildIsPrimarySchoolAge($offsetDate, 0),
```
[the remark this answers is not in the document]
Comment (katjam) [reply app/Services/VoucherEvaluator/EvaluatorFactory.php:67]: Don't we still want to disqualify secondary school children by default as well?

File: database/seeds/SponsorsSeeder.php [hunk 7]
```diff
@@ -11,11 +12,93 @@ class SponsorsSeeder extends Seeder
...
+            new Evaluation([
+                "name" => "FamilyHasUnverifiedChildren",
+                "value" => 0,
+                "purpose" => "notices",
+                "entity" => "App\Family",
+            ])
+        ];
+    }
+
+    public function allowPrimarySchoolers()
```
Comment (katjam) [inline database/seeds/SponsorsSeeder.php:59]: Would these be useful utility functions in for console command? Will be fiddly setting up new sponsors without such tooling.

File: tests/Unit/Services/VoucherEvaluator/VoucherEvaluatorTest.php [hunk 8]
```diff
@@ -18,25 +17,87 @@ class VoucherEvaluatorTest extends TestCase
...
         'ChildIsSecondarySchoolAge' => ['reason' => 'Child|secondary school age'],
         'FamilyHasUnverifiedChildren' => ['reason' => 'Family|has one or more children that you haven\'t checked ID for yet']
     ];
 
     // This has a | in the reason field because we want to carry the entity with it.
     const CREDIT_TYPES = [
-        'ChildIsUnderOne' => ['reason' => 'Child|under 1 year old', 'value' => 3],
-        'ChildIsUnderSchoolAge' => ['reason' => 'Child|under school age', 'value' => 3],
-        'ChildIsUnderTwelve' => ['reason' => 'Child|under 12 years old', 'value' => 3],
-        'ChildIsUnderSecondarySchoolAge' => ['reason' => 'Child|under secondary school age', 'value' => 3],
```
Comment (katjam) [inline tests/Unit/Services/VoucherEvaluator/VoucherEvaluatorTest.php:36]: Was this in use before and no longer `ChildIsSecondarySchoolAge` ?

File: tests/Unit/Services/VoucherEvaluator/VoucherEvaluatorTest.php [hunk 9]
```diff
@@ -18,25 +17,87 @@ class VoucherEvaluatorTest extends TestCase
...
+                new Evaluation([
+                    "name" => "FamilyHasNoEligibleChildren",
+                    "value" => 0,
+                    "purpose" => "disqualifiers",
+                    "entity" => "App\Family",
+                ]),
+            ]
+        );
+
+        //dd($this->rulesMods["credit-extended-qualified"]);
```
Comment (katjam) [inline tests/Unit/Services/VoucherEvaluator/VoucherEvaluatorTest.php:87]: remove?

File: tests/Unit/Services/VoucherEvaluator/VoucherEvaluatorTest.php [hunk 10]
```diff
@@ -90,34 +154,65 @@ public function itCreditsWhenAFamilyIsPregnant()
...
+        $isOverPrimarySchool = factory(Child::class, 'isSecondarySchoolAge')->make();
 
         // Add the kids and check they saved
-        $family->children()->saveMany([$overSchool ,$overSecondarySchool]);
+        $family->children()->saveMany([$isPrimarySchool ,$isOverPrimarySchool]);
+        $this->assertEquals(2, $family->children()->count());
+
+        $evaluation = $evaluator->evaluate($family);
+
+        // Check it can find eligible children (0 vouchers)
```
Comment (katjam) [inline tests/Unit/Services/VoucherEvaluator/VoucherEvaluatorTest.php:178]: I think... The primary kids should get 3 here. And does not matter about the under school age. Alter comment?

File: tests/Unit/Services/VoucherEvaluator/VoucherEvaluatorTest.php [hunk 11]
```diff
@@ -90,34 +154,65 @@ public function itCreditsWhenAFamilyIsPregnant()
...
         // Check it can't find any eligible children (0 vouchers)
-        // - because no under primary school-ers validate the under secondary school-ers.
+        // - because no under primary school-ers validate the primary school-ers.
         $this->assertFalse($evaluation->getEligibility());
         $this->assertEquals('0', $evaluation->getEntitlement());
 
-        // Add a kid that will make the child under secondary school age.
-        $underSchool = factory(Child::class, 'underSchoolAge')->make();
+        // Add a kid that will make the child at primary school age qualified.
+        $underPrimarySchool = factory(Child::class, 'betweenOneAndPrimarySchoolAge')->make();
```
Comment (katjam) [inline tests/Unit/Services/VoucherEvaluator/VoucherEvaluatorTest.php:212]: Is this the right test for extended family rule?
Should it be that having a kid at primary school, under one, or between one and primary will make the at secondary school children qualify?

File: tests/Unit/Services/VoucherEvaluator/VoucherEvaluatorTest.php [hunk 12]
```diff
@@ -170,74 +264,34 @@ public function itCreditsWhenAChildIsUnderSchoolAge()
         $this->assertEquals(1, count($credits));
 
         // Check the correct credit type is applied.
-        $this->assertNotContains(self::CREDIT_TYPES['ChildIsUnderOne'], $credits, '');
-        $this->assertContains(self::CREDIT_TYPES['ChildIsUnderSchoolAge'], $credits, '');
+        $this->assertContains(self::CREDIT_TYPES['ChildIsBetweenOneAndPrimarySchoolAge'], $credits, '');
```
Comment (katjam) [inline tests/Unit/Services/VoucherEvaluator/VoucherEvaluatorTest.php:267]: What's the empty string for?

File: tests/Unit/Services/VoucherEvaluator/VoucherEvaluatorTest.php [hunk 13]
```diff
@@ -170,74 +264,34 @@ public function itCreditsWhenAChildIsUnderSchoolAge()
...
-        // Make a Child under School Age.
-        $child = factory(Child::class, 'underSchoolAge')->make();
+        // Make a Secondary school child
+        $child = factory(Child::class, 'isSecondarySchoolAge')->make();
 
-        // Make standard evaluator
-        $evaluator = EvaluatorFactory::make('extended_age');
-        $evaluation = $evaluator->evaluate($child);
-        $credits = $evaluation["credits"];
+        $rulesMod = collect($this->rulesMods["credit-extended"]);
```
Comment (katjam) [inline tests/Unit/Services/VoucherEvaluator/VoucherEvaluatorTest.php:277]: Do we want credit extended mod here?

File: tests/Unit/Services/VoucherEvaluator/VoucherEvaluatorTest.php [hunk 14]
```diff
@@ -170,74 +264,34 @@ public function itCreditsWhenAChildIsUnderSchoolAge()
...
         // Check the correct credit type is applied.
         $this->assertNotContains(self::CREDIT_TYPES['ChildIsUnderOne'], $credits);
-        $this->assertNotContains(self::CREDIT_TYPES['ChildIsUnderSecondarySchoolAge'], $credits);
+        $this->assertNotContains(self::CREDIT_TYPES['ChildIsBetweenOneAndPrimarySchoolAge'], $credits);
+        $this->assertNotContains(self::CREDIT_TYPES['ChildIsPrimarySchoolAge'], $credits);
+        $this->assertNotContains(self::CREDIT_TYPES['FamilyIsPregnant'], $credits);
         $this->assertEquals(0, $evaluation->getEntitlement());
     }
 
-    // Note, we do not test if a child is overdue or almost born.
```
Comment (katjam) [inline tests/Unit/Services/VoucherEvaluator/VoucherEvaluatorTest.php:238]: remove?

File: tests/Unit/Services/VoucherEvaluator/VoucherEvaluatorTest.php [hunk 15]
```diff
@@ -170,74 +264,34 @@ public function itCreditsWhenAChildIsUnderSchoolAge()
...
+        $evaluator = EvaluatorFactory::make($rulesMod);
         $evaluation = $evaluator->evaluate($child);
         $credits = $evaluation["credits"];
 
-        // Check there's one, because child is not under one.
+        // Check there's none, because child is not primary or under.
         $this->assertEquals(0, count($credits));
 
         // Check the correct credit type is applied.
         $this->assertNotContains(self::CREDIT_TYPES['ChildIsUnderOne'], $credits);
```
Comment (katjam) [inline tests/Unit/Services/VoucherEvaluator/VoucherEvaluatorTest.php:288]: Since there are no credits, are these NotContains relevant to test here?

File: app/Services/VoucherEvaluator/Evaluations/ChildIsBetweenOneAndSecondarySchoolAge.php (hunk 4 above)
Comment (charlesstrange2) [reply app/Services/VoucherEvaluator/Evaluations/ChildIsBetweenOneAndSecondarySchoolAge.php:11]: removed class

File: app/Evaluation.php [hunk 16]
```diff
@@ -0,0 +1,31 @@
...
+        'value',
+        'purpose',
+        'entity',
+        'sponsor_id'
+    ];
+
+    /**
+     * @return \Illuminate\Database\Eloquent\Relations\BelongsTo
+     */
+    public function sponsor()
```
[the remark this answers is not in the document]
Comment (charlesstrange2) [reply app/Evaluation.php:27]: will add

File: app/Services/VoucherEvaluator/Evaluations/ChildIsBetweenOneAndSecondarySchoolAge.php (hunk 4 above)
[the remark this answers is not in the document]
Comment (charlesstrange2) [reply app/Services/VoucherEvaluator/Evaluations/ChildIsBetweenOneAndSecondarySchoolAge.php:11]: nope, will prune

File: app/Services/VoucherEvaluator/EvaluatorFactory.php [hunk 17]
```diff
@@ -51,106 +38,69 @@ public static function make($type = null, Carbon $offsetDate = null)
...
+    public static function generateEvaluations(Collection $modEvaluations, Carbon $offsetDate)
     {
-        return $evaluations = [
-            Child::class => [
+        $evaluations = [
+            "App\Child" => [
                 'credits' => [
-                    new ChildIsUnderOne($offsetDate, 3),
-                    new ChildIsUnderSecondarySchoolAge($offsetDate, 3)
+                    "ChildIsUnderOne" => new ChildIsUnderOne($offsetDate, 6),
```
[the remark this answers is not in the document]
Comment (charlesstrange2) [reply app/Services/VoucherEvaluator/EvaluatorFactory.php:58]: A default has to go somewhere? I could have put it in the evaluations table, i guess, as rules for sponsor_id NULL

File: app/Services/VoucherEvaluator/EvaluatorFactory.php (hunk 6 above)
[the remark this answers is not in the document]
Comment (charlesstrange2) [reply app/Services/VoucherEvaluator/EvaluatorFactory.php:67]: If there's no rules that credit or notice a child, then disqualifying them isn't necessary, they won't get any results

belt and braces we could

- ChildIsOverPrimarySchoolAge
- ChildIsOverSecondarySchoolAge

or even

- ChildIsOverAge($age)

File: app/Services/VoucherEvaluator/EvaluatorFactory.php [hunk 18]
```diff
@@ -51,106 +38,69 @@ public static function make($type = null, Carbon $offsetDate = null)
...
-            Registration::class => [
+            "App\Registration" => [
                 'credits' => [],
                 'notices' => [],
                 'relations' => ['family'],
             ],
         ];
+        $namespace = "App\Services\VoucherEvaluator\Evaluations";
+        // Iterate over the modEvaluations and replace/add them
+        foreach ($modEvaluations as $mod) {
```
[the remark this answers is not in the document]
Comment (charlesstrange2) [reply app/Services/VoucherEvaluator/EvaluatorFactory.php:87]: ok :-)

File: tests/Unit/Services/VoucherEvaluator/VoucherEvaluatorTest.php (hunk 12 above)
Comment (charlesstrange2) [reply tests/Unit/Services/VoucherEvaluator/VoucherEvaluatorTest.php:267]: normally, a message to return; redundant, will remove

File: tests/Unit/Services/VoucherEvaluator/VoucherEvaluatorTest.php (hunk 14 above)
Comment (charlesstrange2) [reply tests/Unit/Services/VoucherEvaluator/VoucherEvaluatorTest.php:238]: those rules got dead-coded

File: tests/Unit/Services/VoucherEvaluator/VoucherEvaluatorTest.php (hunk 15 above)
Comment (charlesstrange2) [reply tests/Unit/Services/VoucherEvaluator/VoucherEvaluatorTest.php:288]: aha! true! will remove

File: tests/Unit/Services/VoucherEvaluator/VoucherEvaluatorTest.php (hunk 13 above)
Comment (charlesstrange2) [reply tests/Unit/Services/VoucherEvaluator/VoucherEvaluatorTest.php:277]: it's a rulesMod element. I think that's ok?

File: tests/Unit/Services/VoucherEvaluator/VoucherEvaluatorTest.php (hunk 11 above)
Comment (charlesstrange2) [reply tests/Unit/Services/VoucherEvaluator/VoucherEvaluatorTest.php:212]: No, secondary schoolers don't get vouchers yet. 

normally
- pregnancy - 3
- under one - 6
- toddler - 3
- primary - 0

the credit-primary changes that to
- primary - 3

the credit-primary-qualified qualfies that with
- primary (with a toddler, under one,  or pregnancy) - 3

**Comment by charlesstrange2** [comment]
Hmm. Github doesn't appear to be talking to gitlab

File: app/Services/VoucherEvaluator/EvaluatorFactory.php (hunk 17 above)
[the remark this answers is not in the document]
Comment (katjam) [reply app/Services/VoucherEvaluator/EvaluatorFactory.php:58]: That's interesting idea. But Maybe also confusing? I was thinking maybe by making them CONSTs in one place - even in this file.

**Comment by charlesstrange2** [comment]
> A few comments but this doesn't look in too bad shape. Nice job!
> 
> Do we want to include a migration to configure the existing sponsors, or are we going with defaults and only have Evaluations for areas that are not following the default.
> 
> Also - we missed out the display config in Admin portal. I haven't checked the trello, but I guess there are more outstanding tasks?

we did miss the admin console :-( but they know that

**Comment by charlesstrange2** [comment]
Forcing because the github-travis link appears to have lost it's marbles, but it passes
https://travis-ci.org/github/neontribe/ARCVService/builds/692122765

File: app/Services/VoucherEvaluator/Evaluations/FamilyHasNoEligibleChildren.php [hunk 19]
```diff
@@ -16,10 +16,11 @@ class FamilyHasNoEligibleChildren extends BaseFamilyEvaluation
     /**
      * FamilyIsPregnant constructor.
```
Comment (Nikomus) [inline app/Services/VoucherEvaluator/Evaluations/FamilyHasNoEligibleChildren.php:17]: Different class.. 🕵️

**Comment by charlesstrange2** [comment]
Bugger!

tech debt at this point, alas.

**Comment by charlesstrange2** [comment]
good spot though @Nikomus , you found my copy/pasta! 🍝
